src/client/serialize/daos_serialize.c

```diff
@@ -1175,3 +1252,1807 @@ daos_cont_deserialize_attrs(char *filename, uint64_t *num_attrs,
+		usr_attr_memtype = H5Tcreate(H5T_COMPOUND,
+					     sizeof(struct usr_attr));
+		if (usr_attr_memtype < 0) {
+			D_ERROR("failed to create memory datatype\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+
+		usr_attr_name_vtype = H5Tcopy(H5T_C_S1);
+		if (usr_attr_name_vtype < 0) {
+			D_ERROR("failed to create variable datatype\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+		status = H5Tset_size(usr_attr_name_vtype, H5T_VARIABLE);
+		if (status < 0) {
+			D_ERROR("failed to set datatype size\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+		usr_attr_val_vtype = H5Tvlen_create(H5T_NATIVE_OPAQUE);
+		if (usr_attr_val_vtype < 0) {
+			D_ERROR("failed to variable length type\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+		status = H5Tinsert(usr_attr_memtype, "Attribute Name",
+				   HOFFSET(struct usr_attr, attr_name),
+				   usr_attr_name_vtype);
+		if (status < 0) {
+			D_ERROR("failed to insert into compound datatype\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+		status = H5Tinsert(usr_attr_memtype, "Attribute Value",
+				   HOFFSET(struct usr_attr, attr_val),
+				   usr_attr_val_vtype);
+		if (status < 0) {
+			D_ERROR("failed to insert into compound datatype\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+		rc = daos_cont_serialize_attrs(hdf5.file, &usr_attr_memtype,
+					       num_attrs, names, buffers,
+					       sizes);
+		if (rc != 0) {
+			D_ERROR("failed to serialize usr attributes "DF_RC"\n",
+				DP_RC(rc));
+			D_GOTO(out, rc);
+		}
+	}
+
+	/* create oid, dkey, akey buffers for container data */
+	D_ALLOC(hdf5.oid_data, sizeof(struct oid));
+	if (hdf5.oid_data == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	D_ALLOC(hdf5.dkey_data, sizeof(struct dkey));
+	if (hdf5.dkey_data == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	D_ALLOC(hdf5.akey_data, sizeof(struct akey));
+	if (hdf5.akey_data == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	hdf5.oid = &(hdf5.oid_data);
+	hdf5.dk = &(hdf5.dkey_data);
+	hdf5.ak = &(hdf5.akey_data);
+	hdf5.oid_buf_size = sizeof(struct oid);
+	hdf5.dkey_buf_size = sizeof(struct dkey);
+	hdf5.akey_buf_size = sizeof(struct akey);
+
+	/* create snapshot, open oit, then starting iterating over oids */
+	rc = daos_cont_create_snap_opt(coh, &epoch, NULL,
+				       DAOS_SNAP_OPT_CR | DAOS_SNAP_OPT_OIT, NULL);
+	if (rc != 0) {
+		D_ERROR("Failed to create snapshot: "DF_RC, DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+	rc = daos_oit_open(coh, epoch, &toh, NULL);
+	if (rc != 0) {
+		D_ERROR("Failed to open object iterator: "DF_RC, DP_RC(rc));
+		D_GOTO(out_snap, rc);
+	}
+	memset(&anchor, 0, sizeof(anchor));
+	while (!daos_anchor_is_eof(&anchor)) {
+		oids_nr = OID_ARR_SIZE;
+		rc = daos_oit_list(toh, oids, &oids_nr, &anchor, NULL);
+		if (rc != 0) {
+			D_ERROR("Failed to list objects: "DF_RC, DP_RC(rc));
+			D_GOTO(out_oit, rc);
+		}
+
+		hdf5.oid_old_buf_size = hdf5.oid_buf_size;
+		hdf5.oid_buf_size = (oids_nr + *total_oids) * sizeof(struct oid);
+		D_REALLOC(oid_tmp, hdf5.oid_data, hdf5.oid_old_buf_size, hdf5.oid_buf_size);
```

Comment:
maybe it's not a good idea to keep the entire oid array in memory.
there could be billions of oids. so why not store the oids in the dataset a bunch at a time instead all at once?